### PR TITLE
Revert breaking changes around BotFrameworkClient implementations

### PR DESCRIPTION
**Pull request by stevengum**
## Description
Fixes breaking changes caused by adding `<T>` to BotframeworkClient and its subclasses
Adds overloads to BotFrameworkHttpClient and SkillHttpClient

### **BREAKING CHANGES: Those extending either of the BotFrameworkHttpClient or the SkillClient will need to implement the new templated overload.**

**Those using the base implementation of the class should not have to change any code.** 

____

## Specific Changes
  - **BotFrameworkClient is now an interface with a Union-type for the postActivity() property.**
    - The only difference in the union type is that one of the subtypes takes a template (`T`) for its return value
     - The implementation of the interface would look like this: `postActivity<T>(...): Promise<InvokeResponse<T>>`
     - The union type is for enabling future implementations of BotFrameworkClient to not have to implement an unused abstract method while not breaking current implementations of BotFrameworkClient
  - Fix bug in `runDialog()`

## Testing

<details>
<summary>4.7.2 BotFrameworkHttpClient typings</summary>

There was no BotFrameworkClient interface so the type below is the current type in usage.

```ts
export declare class BotFrameworkHttpClient {
    private readonly credentialProvider;
    private readonly channelService?;
    /**
     * Cache for appCredentials to speed up token acquisition (a token is not requested unless is expired)
     * AppCredentials are cached using appId + scope (this last parameter is only used if the app credentials are used to call a skill)
     */
    private static readonly appCredentialMapCache;
    constructor(credentialProvider: ICredentialProvider, channelService?: string);
    /**
     * Forwards an activity to a another bot.
     * @remarks
     *
     * @param fromBotId The MicrosoftAppId of the bot sending the activity.
     * @param toBotId The MicrosoftAppId of the bot receiving the activity.
     * @param toUrl The URL of the bot receiving the activity.
     * @param serviceUrl The callback Url for the skill host.
     * @param conversationId A conversation ID to use for the conversation with the skill.
     * @param activity Activity to forward.
     */
    postActivity(fromBotId: string, toBotId: string, toUrl: string, serviceUrl: string, conversationId: string, activity: Activity): Promise<InvokeResponse>;
    /**
     * Gets the application credentials. App Credentials are cached so as to ensure we are not refreshing
     * token every time.
     * @private
     * @param appId The application identifier (AAD Id for the bot).
     * @param oAuthScope The scope for the token, skills will use the Skill App Id.
     */
    private getAppCredentials;
}
```
</details>

<details>
<summary><i>4.8.0-rc0's breaking changes BotFrameworkClient typings</i></summary>

The problem below is in the addition of the Template `T`

```ts
export declare abstract class BotFrameworkClient {
    /**
     * Forwards an activity to a another bot.
     * @remarks
     *
     * @param fromBotId The MicrosoftAppId of the bot sending the activity.
     * @param toBotId The MicrosoftAppId of the bot receiving the activity.
     * @param toUrl The URL of the bot receiving the activity.
     * @param serviceUrl The callback Url for the skill host.
     * @param conversationId A conversation ID to use for the conversation with the skill.
     * @param activity Activity to forward.
     */
    abstract postActivity<T>(fromBotId: string, toBotId: string, toUrl: string, serviceUrl: string, conversationId: string, activity: Activity): Promise<InvokeResponse<T>>;
}
```
</details>

<details>
<summary><strong>PR's BotFrameworkClient typings</strong></summary>

```ts
export interface BotFrameworkClient {
    /**
     * Forwards an activity to a another bot.
     * @remarks
     *
     * @param fromBotId The MicrosoftAppId of the bot sending the activity.
     * @param toBotId The MicrosoftAppId of the bot receiving the activity.
     * @param toUrl The URL of the bot receiving the activity.
     * @param serviceUrl The callback Url for the skill host.
     * @param conversationId A conversation ID to use for the conversation with the skill.
     * @param activity Activity to forward.
     */
    postActivity: (<T>(fromBotId: string, toBotId: string, toUrl: string, serviceUrl: string, conversationId: string, activity: Activity) => Promise<InvokeResponse<T>>) | ((fromBotId: string, toBotId: string, toUrl: string, serviceUrl: string, conversationId: string, activity: Activity) => Promise<InvokeResponse>);
}
```
</details>

<details>
<summary><strong>PR's BotFrameworkHttpClient typings</strong></summary>

```ts
export declare class BotFrameworkHttpClient implements BotFrameworkClient {
    protected readonly channelService: string;
    /**
     * Cache for appCredentials to speed up token acquisition (a token is not requested unless is expired)
     * AppCredentials are cached using appId + scope (this last parameter is only used if the app credentials are used to call a skill)
     */
    private static readonly appCredentialMapCache;
    private readonly credentialProvider;
    constructor(credentialProvider: ICredentialProvider, channelService?: string);
    /**
     * Forwards an activity to a another bot.
     * @remarks
     *
     * @template T The type of body in the InvokeResponse.
     * @param fromBotId The MicrosoftAppId of the bot sending the activity.
     * @param toBotId The MicrosoftAppId of the bot receiving the activity.
     * @param toUrl The URL of the bot receiving the activity.
     * @param serviceUrl The callback Url for the skill host.
     * @param conversationId A conversation ID to use for the conversation with the skill.
     * @param activity Activity to forward.
     */
    postActivity<T>(fromBotId: string, toBotId: string, toUrl: string, serviceUrl: string, conversationId: string, activity: Activity): Promise<InvokeResponse<T>>;
    postActivity(fromBotId: string, toBotId: string, toUrl: string, serviceUrl: string, conversationId: string, activity: Activity): Promise<InvokeResponse>;
    protected buildCredentials(appId: string, oAuthScope?: string): Promise<AppCredentials>;
    /**
     * Gets the application credentials. App Credentials are cached so as to ensure we are not refreshing
     * token every time.
     * @private
     * @param appId The application identifier (AAD Id for the bot).
     * @param oAuthScope The scope for the token, skills will use the Skill App Id.
     */
    private getAppCredentials;
}
```
</details>

## SkillHttpClient typings

<details>
<summary>4.7.2 SkillHttpClient typings</summary>

```ts
export declare class SkillHttpClient extends BotFrameworkHttpClient {
    private readonly conversationIdFactory;
    constructor(credentialProvider: ICredentialProvider, conversationIdFactory: SkillConversationIdFactoryBase, channelService?: string);
    postToSkill(fromBotId: string, toSkill: BotFrameworkSkill, serviceUrl: string, activity: Activity): Promise<InvokeResponse>;
}
```
</details>

<details>
<summary><i>4.8.0-rc0's breaking changes SkillHttpClient typings</i></summary>

The problem below is in the addition of the Template `T`

```ts
export declare class SkillHttpClient extends BotFrameworkHttpClient {
    private readonly conversationIdFactory;
    constructor(credentialProvider: ICredentialProvider, conversationIdFactory: SkillConversationIdFactoryBase, channelService?: string);
    /**
     *
     * @param originatingAudience
     * @param fromBotId
     * @param toSkill
     * @param callbackUrl
     * @param activity
     */
    postToSkill<T>(originatingAudience: string, fromBotId: string, toSkill: BotFrameworkSkill, callbackUrl: string, activity: Activity): Promise<InvokeResponse<T>>;
    postToSkill<T>(fromBotId: string, toSkill: BotFrameworkSkill, callbackUrl: string, activity: Activity): Promise<InvokeResponse>;
}
```
</details>

<details>
<summary><strong>PR's SkillHttpClient typings</strong></summary>

```ts
export declare class SkillHttpClient extends BotFrameworkHttpClient {
    private readonly conversationIdFactory;
    constructor(credentialProvider: ICredentialProvider, conversationIdFactory: SkillConversationIdFactoryBase, channelService?: string);
    /**
     * Uses the SkillConversationIdFactory to create or retrieve a Skill Conversation Id, and sends the activity.
     * @template T The type of body in the InvokeResponse.
     * @param originatingAudience The OAuth audience scope, used during token retrieval. (Either https://api.botframework.com or bot app id.)
     * @param fromBotId The MicrosoftAppId of the bot sending the activity.
     * @param toSkill The skill to create the Conversation Id for.
     * @param callbackUrl The callback Url for the skill host.
     * @param activity The activity to send.
     */
    postToSkill<T>(originatingAudience: string, fromBotId: string, toSkill: BotFrameworkSkill, callbackUrl: string, activity: Activity): Promise<InvokeResponse<T>>;
    /**
     * Uses the SkillConversationIdFactory to create or retrieve a Skill Conversation Id, and sends the activity.
     * @deprecated This overload is deprecated. Please use SkillHttpClient.postToSkill() that takes an `originatingAudience`.
     * @param fromBotId The MicrosoftAppId of the bot sending the activity.
     * @param toSkill The skill to create the Conversation Id for.
     * @param callbackUrl The callback Url for the skill host.
     * @param activity The activity to send.
     */
    postToSkill(fromBotId: string, toSkill: BotFrameworkSkill, callbackUrl: string, activity: Activity): Promise<InvokeResponse>;
}
```
</details>

### CODE REVIEW

**Comment by stevengum** [comment]
Waiting for @christopheranderson's review

File: libraries/botbuilder/src/botFrameworkHttpClient.ts [hunk 1]
```diff
@@ -116,7 +123,7 @@ export class BotFrameworkHttpClient extends BotFrameworkClient {
         }
     }
 
-    protected async buildCredentials(appId: string, oAuthScope?: string): Promise<MicrosoftAppCredentials> {
+    protected async buildCredentials(appId: string, oAuthScope?: string): Promise<AppCredentials> {
```
Comment (christopheranderson) [inline libraries/botbuilder/src/botFrameworkHttpClient.ts]: Just confirming that 4.7 doesn't have a buildCredentials method with MicrosoftAPpCredentials as a return type. Protected is still public for folks who extend the API.

Comment (stevengum) [reply libraries/botbuilder/src/botFrameworkHttpClient.ts]: Technically this is loosening the the API as MicrosoftAppCredentials subclasses AppCredentials... 

But this technically is a breaking change if users now have to cast the return value from their code... Thoughts?

Comment (stevengum) [reply libraries/botbuilder/src/botFrameworkHttpClient.ts]: Additional caveat, this makes the signature the same as C#: [link](https://github.com/microsoft/botbuilder-dotnet/blob/2ebf6d118ceebece83ff7b5b38edbc2d8a253679/libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotFrameworkHttpClient.cs#L146)

**Pull request merged**